server/routes/serviceProviderReferrals/interventionProgressPresenter.ts

```diff
@@ -53,18 +55,48 @@ export default class InterventionProgressPresenter {
     }
 
     return this.actionPlanAppointments.map(appointment => {
-      const editUrl = `/service-provider/action-plan/${this.actionPlan!.id}/sessions/${appointment.sessionNumber}/edit`
+      const sessionTableParams = this.sessionTableParams(appointment)
 
       return {
         sessionNumber: appointment.sessionNumber,
         appointmentTime: DateUtils.formatDateTimeOrEmptyString(appointment.appointmentTime),
-        tagArgs: appointment.appointmentTime
-          ? { text: 'SCHEDULED', classes: 'govuk-tag--blue' }
-          : { text: 'NOT SCHEDULED', classes: 'govuk-tag--grey' },
-        linkHtml: appointment.appointmentTime
-          ? `<a class="govuk-link" href="${editUrl}">Reschedule session</a><br><a class="govuk-link" href="/service-provider/action-plan/${this.actionPlan?.id}/appointment/${appointment.sessionNumber}/post-session-feedback/attendance">Give feedback</a>`
-          : `<a class="govuk-link" href="${editUrl}">Edit session details</a>`,
+        tagArgs: { text: sessionTableParams.text, classes: sessionTableParams.tagClass },
+        linkHtml: sessionTableParams.linkHTML,
       }
     })
   }
+
+  private sessionTableParams(appointment: ActionPlanAppointment): { text: string; tagClass: string; linkHTML: string } {
+    const status = sessionStatus.forAppointment(appointment)
+    const presenter = new SessionStatusPresenter(status)
+
+    const editUrl = `/service-provider/action-plan/${this.actionPlan!.id}/sessions/${appointment.sessionNumber}/edit`
+
+    switch (status) {
+      case SessionStatus.didNotAttend:
+        return {
+          text: presenter.text,
```

Comment:
Did the `{ ...presenter, linkHTML: 'blah blah' }` suggestion not work here? It would be more succinct.

Reply:
I get `Type '{ linkHTML: string; }' is missing the following properties from type '{ text: string; tagClass: string; linkHTML: string; }': text, tagClass` - it seems as though it's not picking up either of the fields.

Interestingly, it doesn't complain if I do:
```
const { text, tagClass } = new SessionStatusPresenter(status)
{
  ...{ text, tagClass },
  linkHTML: `<a class="govuk-link" href="#">View feedback form</a>`,
}
```

Do you think that's at all cleaner?


My theory _was_ that this was happening because `new SessionStatusPresenter(status)` doesn't return an object literal e.g. `{ text: "", tagClass: ""}`, it returns some other type of object, but that doesn't explain why the above is acceptable?

Reply:
It seems to be because the presenter is using getters to implement the `text` and `tagClass` properties, and it seems like the spread operator (`...`) doesn't gather properties that are defined by getters. Changing the presenter implementation to use normal properties (see example below) makes it compile, but I don't think it matters.

```typescript
export default class SessionStatusPresenter {
  constructor(private readonly status: SessionStatus) {}

  readonly text = (() => {
    switch (this.status) {
      case SessionStatus.didNotAttend:
        return 'did not attend'
      case SessionStatus.scheduled:
        return 'scheduled'
      case SessionStatus.completed:
        return 'completed'
      default:
        return 'not scheduled'
    }
  })()

  readonly tagClass = (() => {
    switch (this.status) {
      case SessionStatus.didNotAttend:
        return 'govuk-tag--purple'
      case SessionStatus.scheduled:
        return 'govuk-tag--blue'
      case SessionStatus.completed:
        return 'govuk-tag--green'
      default:
        return 'govuk-tag--grey'
    }
  })()
}
```